sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/throughputControl/controller/group/ThroughputGroupControllerBase.java

```diff
@@ -0,0 +1,238 @@
+    private final AtomicReference<Double> groupThroughput;
+    private final AtomicInteger maxContainerThroughput;
+    private final RxPartitionKeyRangeCache partitionKeyRangeCache;
+    private final AsyncCache<String, IThroughputRequestController> requestControllerAsyncCache;
+    private final String targetContainerRid;
+
+    private final CancellationTokenSource cancellationTokenSource;
+    private final Scheduler scheduler;
+
+    public ThroughputGroupControllerBase(
+        ConnectionMode connectionMode,
+        GlobalEndpointManager globalEndpointManager,
+        ThroughputControlGroup group,
+        Integer maxContainerThroughput,
+        RxPartitionKeyRangeCache partitionKeyRangeCache,
+        String targetContainerRid) {
+
+        checkNotNull(globalEndpointManager, "Global endpoint manager can not be null");
+        checkNotNull(group, "Throughput control group can not be null");
+        checkNotNull(partitionKeyRangeCache, "Partition key range cache can not be null or empty");
+        checkArgument(StringUtils.isNotEmpty(targetContainerRid), "Target container rid cannot be null nor empty");
+
+        this.connectionMode = connectionMode;
+        this.globalEndpointManager = globalEndpointManager;
+        this.group = group;
+
+        if (this.group.getTargetThroughputThreshold() != null) {
+            checkNotNull(maxContainerThroughput, "Max container throughput can not be null when target throughput threshold defined");
+            this.maxContainerThroughput = new AtomicInteger(maxContainerThroughput);
+        } else {
+            this.maxContainerThroughput = null;
+        }
+
+        this.groupThroughput = new AtomicReference<>();
+        this.calculateGroupThroughput();// Calculate after maxContainerThroughput
+
+        this.partitionKeyRangeCache = partitionKeyRangeCache;
+        this.requestControllerAsyncCache = new AsyncCache<>();
+        this.targetContainerRid = targetContainerRid;
+
+        this.cancellationTokenSource = new CancellationTokenSource();
+        this.scheduler = Schedulers.elastic();
+    }
+
+    private void calculateGroupThroughput() {
+        double allocatedThroughput = Double.MAX_VALUE;
+        if (this.group.getTargetThroughputThreshold() != null) {
+            allocatedThroughput = Math.min(allocatedThroughput, this.maxContainerThroughput.get() * this.group.getTargetThroughputThreshold());
+        }
+
+        if (this.group.getTargetThroughput() != null) {
+            allocatedThroughput = Math.min(allocatedThroughput, this.group.getTargetThroughput());
+        }
+
+        this.groupThroughput.set(allocatedThroughput);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T> Mono<T> init() {
+        return this.resolveRequestController()
+            .doOnSuccess(dummy -> {
+                scheduler.schedule(() -> this.throughputUsageCycleRenewTask(this.cancellationTokenSource.getToken()).subscribe());
```

Comment:
Subscribing here can have diverse effects, since we are not sure when this init will complete. We need to discuss this offline on the correct usage of this.

Reply:
Ack. will update in the next iteration.

Reply:
Need to start the task when the request controller initialized successfully. Will keep it for now.